crypto/provider_conf.c

```diff
@@ -0,0 +1,170 @@
+                                const CONF *cnf)
+{
+    STACK_OF(CONF_VALUE) *sect;
+
+    OSSL_TRACE2(PROVIDER_CONF, "PROVIDER conf: %s = %s\n", name, value);
+
+    sect = NCONF_get_section(cnf, value);
+    if (sect != NULL) {
+        int i;
+        char buffer[512];
+        size_t buffer_len = 0;
+
+        if (name != NULL) {
+            OPENSSL_strlcpy(buffer, name, sizeof(buffer));
+            OPENSSL_strlcat(buffer, ".", sizeof(buffer));
+            buffer_len = strlen(buffer);
+        }
+
+        for (i = 0; i < sk_CONF_VALUE_num(sect); i++) {
+            CONF_VALUE *sectconf = sk_CONF_VALUE_value(sect, i);
+
+            if (buffer_len + strlen(sectconf->name) >= sizeof(buffer))
+                return 0;
+            buffer[buffer_len] = '\0';
+            OPENSSL_strlcat(buffer, sectconf->name, sizeof(buffer));
+            if (!provider_conf_params(prov, buffer, sectconf->value, cnf))
+                return 0;
+        }
+    } else {
+        ossl_provider_add_parameter(prov, name, value);
+    }
+
+    return 1;
+}
+
+static OSSL_PROVIDER *provider_conf_load(OPENSSL_CTX *libctx,
+                                         const char *name, const char *value,
+                                         const CONF *cnf)
+{
+    int i;
+    STACK_OF(CONF_VALUE) *ecmds;
+    int soft = 0;
+    OSSL_PROVIDER *prov = NULL;
+    const char *path = NULL;
+    long activate = 0;
+
+    name = skip_dot(name);
+    OSSL_TRACE1(PROVIDER_CONF, "Configuring provider %s\n", name);
+    /* Value is a section containing PROVIDER commands */
+    ecmds = NCONF_get_section(cnf, value);
+
+    if (!ecmds) {
+        CRYPTOerr(CRYPTO_F_PROVIDER_CONF_LOAD, CRYPTO_R_PROVIDER_SECTION_ERROR);
+        return 0;
+    }
+
+    /* Find the needed data first */
+    for (i = 0; i < sk_CONF_VALUE_num(ecmds); i++) {
+        CONF_VALUE *ecmd = sk_CONF_VALUE_value(ecmds, i);
+        const char *confname = skip_dot(ecmd->name);
+        const char *confvalue = ecmd->value;
+
+        OSSL_TRACE2(PROVIDER_CONF, "PROVIDER conf: %s = %s\n",
+                    confname, confvalue);
+
+        /* First handle some special pseudo confs */
+
+        /* Override provider name to use */
+        if (strcmp(confname, "identity") == 0)
+            name = confvalue;
+        else if (strcmp(confname, "soft_load") == 0)
+            soft = 1;
+        /* Load a dynamic PROVIDER */
+        else if (strcmp(confname, "module") == 0)
+            path = confvalue;
+        else if (strcmp(confname, "activate") == 0)
+            activate = 1;
+    }
+
+    prov = ossl_provider_new(libctx, name, NULL);
+    if (prov == NULL) {
+        if (soft)
+            ERR_clear_error();
+        return NULL;
+    }
+
+    if (path != NULL)
+        ossl_provider_set_module_path(prov, path);
+
+    provider_conf_params(prov, NULL, value, cnf);
+
+    if (activate) {
+        ossl_provider_activate(prov);
+        if (activated_providers == NULL)
```

Comment:
Is activated_providers thread safe?

Reply:
This is all governed by `OPENSSL_init_crypto`, is thread safety in config modules relevant?  If it is, we'd better review *all* our config modules.